eth2/fork_choice/src/longest_chain.rs

```diff
@@ -1,93 +1,102 @@
 }
 
-pub fn longest_chain<T>(
-    head_block_hashes: &[Hash256],
-    block_store: &Arc<BeaconBlockStore<T>>,
-) -> Result<Option<usize>, ForkChoiceError>
+impl<T> LongestChain<T>
 where
     T: ClientDB + Sized,
 {
-    let mut head_blocks: Vec<(usize, BeaconBlock)> = vec![];
-
-    /*
-     * Load all the head_block hashes from the DB as SszBeaconBlocks.
-     */
-    for (index, block_hash) in head_block_hashes.iter().enumerate() {
-        let ssz = block_store
-            .get(&block_hash)?
-            .ok_or(ForkChoiceError::MissingBlock)?;
-        let (block, _) = BeaconBlock::ssz_decode(&ssz, 0)?;
-        head_blocks.push((index, block));
+    pub fn new(block_store: Arc<BeaconBlockStore<T>>) -> Self {
+        LongestChain {
+            head_block_hashes: Vec::new(),
+            block_store,
+        }
     }
+}
 
-    /*
-     * Loop through all the head blocks and find the highest slot.
-     */
-    let highest_slot: Option<Slot> = None;
-    for (_, block) in &head_blocks {
-        let slot = block.slot;
+impl<T: ClientDB + Sized> ForkChoice for LongestChain<T> {
+    fn add_block(
+        &mut self,
+        block: &BeaconBlock,
+        block_hash: &Hash256,
+    ) -> Result<(), ForkChoiceError> {
+        // add the block hash to head_block_hashes removing the parent if it exists
+        self.head_block_hashes
+            .retain(|hash| *hash != block.parent_root);
+        self.head_block_hashes.push(*block_hash);
+        Ok(())
+    }
 
-        match highest_slot {
-            None => Some(slot),
-            Some(winning_slot) => {
-                if slot > winning_slot {
-                    Some(slot)
-                } else {
-                    Some(winning_slot)
-                }
-            }
-        };
+    fn add_attestation(&mut self, _: u64, _: &Hash256) -> Result<(), ForkChoiceError> {
+        // do nothing
+        Ok(())
     }
 
-    /*
-     * Loop through all the highest blocks and sort them by highest hash.
-     *
-     * Ultimately, the index of the head_block hash with the highest slot and highest block
-     * hash will be the winner.
-     */
-    match highest_slot {
-        None => Ok(None),
-        Some(highest_slot) => {
-            let mut highest_blocks = vec![];
-            for (index, block) in head_blocks {
-                if block.slot == highest_slot {
-                    highest_blocks.push((index, block))
-                }
-            }
+    fn find_head(&mut self, _: &Hash256) -> Result<Hash256, ForkChoiceError> {
+        let mut head_blocks: Vec<(usize, BeaconBlock)> = vec![];
+        /*
+         * Load all the head_block hashes from the DB as SszBeaconBlocks.
+         */
+        for (index, block_hash) in self.head_block_hashes.iter().enumerate() {
+            let block = self
+                .block_store
+                .get_deserialized(&block_hash)?
+                .ok_or(ForkChoiceError::MissingBeaconBlock(*block_hash))?;
```

Comment:
Clippy wants this to be `ok_or_else`

Reply:
Yeah, I usually ok_or_else these, this one slipped though. will fix.